### ISSUE TITLE
WebSocket throws a vague IOException when internet is disconnected

### ISSUE THREAD

**Issue by LegendOfLelouch** [ktorio/ktor#1119]
### Ktor Version
1.2.0-rc
### Ktor Engine Used(client or server and name)
Client/CIO
### JVM Version, Operating System and Relevant Context
OpenJDK 11.0.3, Arch Linux.
### Feedback
When the internet is disconnected, a vague `java.io.IOException` is thrown twice, and it cannot be caught.

```
Exception in thread "pool-2-thread-3" java.io.IOException: Connection reset by peer
	at java.base/sun.nio.ch.FileDispatcherImpl.read0(Native Method)
	at java.base/sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
	at java.base/sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:276)
	at java.base/sun.nio.ch.IOUtil.read(IOUtil.java:233)
	at java.base/sun.nio.ch.IOUtil.read(IOUtil.java:223)
	at java.base/sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:358)
	at kotlinx.io.nio.ChannelsKt.read(Channels.kt:117)
	at io.ktor.network.sockets.CIOReaderKt$attachForReadingDirectImpl$1$1.invokeSuspend(CIOReader.kt:75)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.ResumeModeKt.resumeMode(ResumeMode.kt:67)
	at kotlinx.coroutines.DispatchedKt.resume(Dispatched.kt:309)
	at kotlinx.coroutines.DispatchedKt.resumeUnconfined(Dispatched.kt:49)
	at kotlinx.coroutines.DispatchedKt.dispatch(Dispatched.kt:295)
	at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:250)
	at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:260)
	at kotlinx.coroutines.CancellableContinuationImpl.resumeWith(CancellableContinuationImpl.kt:189)
	at io.ktor.network.selector.SelectorManagerSupport.handleSelectedKey(SelectorManagerSupport.kt:83)
	at io.ktor.network.selector.SelectorManagerSupport.handleSelectedKeys(SelectorManagerSupport.kt:63)
	at io.ktor.network.selector.ActorSelectorManager.process(ActorSelectorManager.kt:74)
	at io.ktor.network.selector.ActorSelectorManager$process$1.invokeSuspend(ActorSelectorManager.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:238)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by rocketraman** [comment]
This is not limited to websockets, nor is it limited to network disconnection. I have this situation for a straight persistent HTTP CIO connection, and it happens when the remote server disconnects the socket. The exception output is distracting and unnecessary, as there is nothing the client code can do about it.

**Comment by ratiofu** [comment]
Looks like a duplicate of #545, so this has been an issue for a while.

**Comment by ratiofu** [comment]
@rocketraman I am using a global exception handler as a work-around, but it's not ideal; code fragments:

```kotlin
...
class UncaughtExceptionHandler : Thread.UncaughtExceptionHandler {
  
  override fun uncaughtException(thread: Thread?, throwable: Throwable?) {
    when {
      
      throwable is OutOfMemoryError ->
        try {
          log.error("out of memory", throwable)
        } catch (throwable: Throwable) {
          System.err.println("error attempting to log OOM error:")
          System.err.println(throwable.message)
        }
      
      // handling unexpected connection loss deep in Ktor/Java's IO stack
      throwable is IOException && throwable.message == "Connection reset by peer" ->
        log.info("client connection closed")
      
      else -> log.warn("uncaught error", throwable)
    }
  }
...
```
Somewhere early in the app's startup code:

```kotlin
val defaultExceptionHandler = UncaughtExceptionHandler()
Thread.currentThread().uncaughtExceptionHandler = defaultExceptionHandler
Thread.setDefaultUncaughtExceptionHandler(defaultExceptionHandler)
```

**Comment by brendanw** [comment]
Is this exception supposed to be thrown when the internet is disconnected? I am seeing this exception on a benchmarking test when attempting to make 100,000 concurrent requests.

```
Exception in thread "main" java.net.SocketException: Connection reset by peer
        at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
        at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717)
        at io.ktor.network.sockets.SocketImpl.connect$ktor_network(SocketImpl.kt:34)
        at io.ktor.network.sockets.SocketImpl$connect$1.invokeSuspend(SocketImpl.kt)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

https://github.com/brendanw/profilehttp

Are clients expected to handle this error? It would be great to see some guidance from the ktor team.

**Comment by oleg-larshin** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/KTOR-623">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.

**Comment by e5l** [comment]
Fixed since 1.4.0
